File: lib/src/embedded/boundings/style/application/codec.dart
part of mapbox_navigation_flutter;

/// A custom codec for encoding and decoding messages related to the StyleManager class.
///
/// This codec extends the StandardMessageCodec class and provides the necessary
/// functionality to encode and decode messages specific to the StyleManager class.
class StyleAPICodec extends StandardMessageCodec {
  /// A codec for managing styles.
  ///
  /// This codec is responsible for managing the styles used in the application.
  /// It provides methods for encoding and decoding style data.
  const StyleAPICodec();
  @override
  void writeValue(WriteBuffer buffer, Object? value) {
    try {
      if (value is CameraBounds) {
        buffer.putUint8(128);
        writeValue(buffer, value.encode());
      } else if (value is CameraBoundsOptions) {
        buffer.putUint8(129);
        writeValue(buffer, value.encode());
      } else if (value is CameraOptions) {
        buffer.putUint8(130);
        writeValue(buffer, value.encode());
      } else if (value is CameraState) {
        buffer.putUint8(131);
        writeValue(buffer, value.encode());
      } else if (value is CanonicalTileID) {
        buffer.putUint8(132);
        writeValue(buffer, value.encode());
      } else if (value is CoordinateBounds) {
        buffer.putUint8(133);
        writeValue(buffer, value.encode());
      } else if (value is CoordinateBoundsZoom) {
        buffer.putUint8(134);
        writeValue(buffer, value.encode());
      } else if (value is FeatureExtensionValue) {
        buffer.putUint8(135);
        writeValue(buffer, value.encode());
      } else if (value is GlyphsRasterizationOptions) {
        buffer.putUint8(136);
        writeValue(buffer, value.encode());
      } else if (value is ImageContent) {
        buffer.putUint8(137);
        writeValue(buffer, value.encode());
      } else if (value is ImageStretches) {
        buffer.putUint8(138);
        writeValue(buffer, value.encode());
      } else if (value is LayerPosition) {
        buffer.putUint8(139);
        writeValue(buffer, value.encode());
      } else if (value is MapAnimationOptions) {
        buffer.putUint8(140);
        writeValue(buffer, value.encode());
      } else if (value is MapDebugOptions) {
        buffer.putUint8(141);
        writeValue(buffer, value.encode());
      } else if (value is MapMemoryBudgetInMegabytes) {
        buffer.putUint8(142);
        writeValue(buffer, value.encode());
      } else if (value is MapMemoryBudgetInTiles) {
        buffer.putUint8(143);
        writeValue(buffer, value.encode());
      } else if (value is MapOptions) {
        buffer.putUint8(144);
        writeValue(buffer, value.encode());
      } else if (value is MbxEdgeInsets) {
        buffer.putUint8(145);
        writeValue(buffer, value.encode());
      } else if (value is MbxImage) {
        buffer.putUint8(146);
        writeValue(buffer, value.encode());
      } else if (value is MercatorCoordinate) {
        buffer.putUint8(147);
        writeValue(buffer, value.encode());
      } else if (value is OfflineRegionGeometryDefinition) {
        buffer.putUint8(148);
        writeValue(buffer, value.encode());
      } else if (value is OfflineRegionTilePyramidDefinition) {
        buffer.putUint8(149);
        writeValue(buffer, value.encode());
      } else if (value is ProjectedMeters) {
        buffer.putUint8(150);
        writeValue(buffer, value.encode());
      } else if (value is QueriedFeature) {
        buffer.putUint8(151);
        writeValue(buffer, value.encode());
      } else if (value is RenderedQueryGeometry) {
        buffer.putUint8(152);
        writeValue(buffer, value.encode());
      } else if (value is RenderedQueryOptions) {
        buffer.putUint8(153);
        writeValue(buffer, value.encode());
      } else if (value is ResourceOptions) {
        buffer.putUint8(154);
        writeValue(buffer, value.encode());
      } else if (value is ScreenBox) {
        buffer.putUint8(155);
        writeValue(buffer, value.encode());
      } else if (value is ScreenCoordinate) {
        buffer.putUint8(156);
        writeValue(buffer, value.encode());
      } else if (value is Size) {
        buffer.putUint8(157);
        writeValue(buffer, value.encode());
      } else if (value is SourceQueryOptions) {
        buffer.putUint8(158);
        writeValue(buffer, value.encode());
      } else if (value is StyleObjectInfo) {
        buffer.putUint8(159);
        writeValue(buffer, value.encode());
      } else if (value is StylePropertyValue) {
        buffer.putUint8(160);
        writeValue(buffer, value.encode());
      } else if (value is TransitionOptions) {
        buffer.putUint8(161);
        writeValue(buffer, value.encode());
      } else {
        super.writeValue(buffer, value);
      }
    } catch (error) {
      print(error);
    }
  }

  @override
  Object? readValueOfType(int type, ReadBuffer buffer) {
    switch (type) {
      case 128:
        return CameraBounds.decode(readValue(buffer)!);
      case 129:
        return CameraBoundsOptions.decode(readValue(buffer)!);
      case 130:
        return CameraOptions.decode(readValue(buffer)!);
      case 131:
        return CameraState.decode(readValue(buffer)!);
      case 132:
        return CanonicalTileID.decode(readValue(buffer)!);
      case 133:
        return CoordinateBounds.decode(readValue(buffer)!);
      case 134:
        return CoordinateBoundsZoom.decode(readValue(buffer)!);
      case 135:
        return FeatureExtensionValue.decode(readValue(buffer)!);
      case 136:
        return GlyphsRasterizationOptions.decode(readValue(buffer)!);
      case 137:
        return ImageContent.decode(readValue(buffer)!);
      case 138:
        return ImageStretches.decode(readValue(buffer)!);
      case 139:
        return LayerPosition.decode(readValue(buffer)!);
      case 140:
        return MapAnimationOptions.decode(readValue(buffer)!);
      case 141:
        return MapDebugOptions.decode(readValue(buffer)!);
      case 142:
        return MapMemoryBudgetInMegabytes.decode(readValue(buffer)!);
      case 143:
        return MapMemoryBudgetInTiles.decode(readValue(buffer)!);
      case 144:
        return MapOptions.decode(readValue(buffer)!);
      case 145:
        return MbxEdgeInsets.decode(readValue(buffer)!);
      case 146:
        return MbxImage.decode(readValue(buffer)!);
      case 147:
        return MercatorCoordinate.decode(readValue(buffer)!);
      case 148:
        return OfflineRegionGeometryDefinition.decode(readValue(buffer)!);
      case 149:
        return OfflineRegionTilePyramidDefinition.decode(readValue(buffer)!);
      case 150:
        return ProjectedMeters.decode(readValue(buffer)!);
      case 151:
        return QueriedFeature.decode(readValue(buffer)!);
      case 152:
        return RenderedQueryGeometry.decode(readValue(buffer)!);
      case 153:
        return RenderedQueryOptions.decode(readValue(buffer)!);
      case 154:
        return ResourceOptions.decode(readValue(buffer)!);
      case 155:
        return ScreenBox.decode(readValue(buffer)!);
      case 156:
        return ScreenCoordinate.decode(readValue(buffer)!);
      case 157:
        return Size.decode(readValue(buffer)!);
      case 158:
        return SourceQueryOptions.decode(readValue(buffer)!);
      case 159:
        return StyleObjectInfo.decode(readValue(buffer)!);
      case 160:
        return StylePropertyValue.decode(readValue(buffer)!);
      case 161:
        return TransitionOptions.decode(readValue(buffer)!);
      default:
        return super.readValueOfType(type, buffer);
    }
  }
}

File: android/src/main/kotlin/com/eopeter/fluttermapboxnavigation/boundings/style/application/Codec.kt
package com.eopeter.fluttermapboxnavigation.boundings.style.application

import com.eopeter.fluttermapboxnavigation.core.*
import com.mapbox.geojson.Point
import io.flutter.plugin.common.StandardMessageCodec
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer

@Suppress("UNCHECKED_CAST")
object StyleApiCodec : StandardMessageCodec() {
    override fun readValueOfType(type: Byte, buffer: ByteBuffer): Any? {
        return when (type) {
            128.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    AmbientLight.fromList(it)
                }
            }
            129.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    CameraBounds.fromList(it)
                }
            }
            130.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    CameraBoundsOptions.fromList(it)
                }
            }
            131.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    CameraOptions.fromList(it)
                }
            }
            132.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    CameraState.fromList(it)
                }
            }
            133.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    CanonicalTileID.fromList(it)
                }
            }
            134.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    CoordinateBounds.fromList(it)
                }
            }
            135.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    CoordinateBoundsZoom.fromList(it)
                }
            }
            136.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    DirectionalLight.fromList(it)
                }
            }
            137.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    FeatureExtensionValue.fromList(it)
                }
            }
            138.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    FlatLight.fromList(it)
                }
            }
            139.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    GlyphsRasterizationOptions.fromList(it)
                }
            }
            140.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    ImageContent.fromList(it)
                }
            }
            141.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    ImageStretches.fromList(it)
                }
            }
            142.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    LayerPosition.fromList(it)
                }
            }
            143.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    MapAnimationOptions.fromList(it)
                }
            }
            144.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    MapDebugOptions.fromList(it)
                }
            }
            145.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    MapOptions.fromList(it)
                }
            }
            146.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    MbxEdgeInsets.fromList(it)
                }
            }
            147.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    MbxImage.fromList(it)
                }
            }
            148.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    MercatorCoordinate.fromList(it)
                }
            }
            149.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    OfflineRegionGeometryDefinition.fromList(it)
                }
            }
            150.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    OfflineRegionTilePyramidDefinition.fromList(it)
                }
            }
            151.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    PointDecoder.fromList(it)
                }
            }
            152.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    ProjectedMeters.fromList(it)
                }
            }
            153.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    QueriedFeature.fromList(it)
                }
            }
            154.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    QueriedRenderedFeature.fromList(it)
                }
            }
            155.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    QueriedSourceFeature.fromList(it)
                }
            }
            156.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    RenderedQueryGeometry.fromList(it)
                }
            }
            157.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    RenderedQueryOptions.fromList(it)
                }
            }
            158.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    ScreenBox.fromList(it)
                }
            }
            159.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    ScreenCoordinate.fromList(it)
                }
            }
            160.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    Size.fromList(it)
                }
            }
            161.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    SourceQueryOptions.fromList(it)
                }
            }
            162.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    StyleObjectInfo.fromList(it)
                }
            }
            163.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    StyleProjection.fromList(it)
                }
            }
            164.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    StylePropertyValue.fromList(it)
                }
            }
            165.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    TileCacheBudgetInMegabytes.fromList(it)
                }
            }
            166.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    TileCacheBudgetInTiles.fromList(it)
                }
            }
            167.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    TransitionOptions.fromList(it)
                }
            }
            else -> super.readValueOfType(type, buffer)
        }
    }

    override fun writeValue(stream: ByteArrayOutputStream, value: Any?) {
        when (value) {
            is AmbientLight -> {
                stream.write(128)
                writeValue(stream, value.toList())
            }
            is CameraBounds -> {
                stream.write(129)
                writeValue(stream, value.toList())
            }
            is CameraBoundsOptions -> {
                stream.write(130)
                writeValue(stream, value.toList())
            }
            is CameraOptions -> {
                stream.write(131)
                writeValue(stream, value.toList())
            }
            is CameraState -> {
                stream.write(132)
                writeValue(stream, value.toList())
            }
            is CanonicalTileID -> {
                stream.write(133)
                writeValue(stream, value.toList())
            }
            is CoordinateBounds -> {
                stream.write(134)
                writeValue(stream, value.toList())
            }
            is CoordinateBoundsZoom -> {
                stream.write(135)
                writeValue(stream, value.toList())
            }
            is DirectionalLight -> {
                stream.write(136)
                writeValue(stream, value.toList())
            }
            is FeatureExtensionValue -> {
                stream.write(137)
                writeValue(stream, value.toList())
            }
            is FlatLight -> {
                stream.write(138)
                writeValue(stream, value.toList())
            }
            is GlyphsRasterizationOptions -> {
                stream.write(139)
                writeValue(stream, value.toList())
            }
            is ImageContent -> {
                stream.write(140)
                writeValue(stream, value.toList())
            }
            is ImageStretches -> {
                stream.write(141)
                writeValue(stream, value.toList())
            }
            is LayerPosition -> {
                stream.write(142)
                writeValue(stream, value.toList())
            }
            is MapAnimationOptions -> {
                stream.write(143)
                writeValue(stream, value.toList())
            }
            is MapDebugOptions -> {
                stream.write(144)
                writeValue(stream, value.toList())
            }
            is MapOptions -> {
                stream.write(145)
                writeValue(stream, value.toList())
            }
            is MbxEdgeInsets -> {
                stream.write(146)
                writeValue(stream, value.toList())
            }
            is MbxImage -> {
                stream.write(147)
                writeValue(stream, value.toList())
            }
            is MercatorCoordinate -> {
                stream.write(148)
                writeValue(stream, value.toList())
            }
            is OfflineRegionGeometryDefinition -> {
                stream.write(149)
                writeValue(stream, value.toList())
            }
            is OfflineRegionTilePyramidDefinition -> {
                stream.write(150)
                writeValue(stream, value.toList())
            }
            is Point -> {
                stream.write(151)
                writeValue(stream, value.toList())
            }
            is ProjectedMeters -> {
                stream.write(152)
                writeValue(stream, value.toList())
            }
            is QueriedFeature -> {
                stream.write(153)
                writeValue(stream, value.toList())
            }
            is QueriedRenderedFeature -> {
                stream.write(154)
                writeValue(stream, value.toList())
            }
            is QueriedSourceFeature -> {
                stream.write(155)
                writeValue(stream, value.toList())
            }
            is RenderedQueryGeometry -> {
                stream.write(156)
                writeValue(stream, value.toList())
            }
            is RenderedQueryOptions -> {
                stream.write(157)
                writeValue(stream, value.toList())
            }
            is ScreenBox -> {
                stream.write(158)
                writeValue(stream, value.toList())
            }
            is ScreenCoordinate -> {
                stream.write(159)
                writeValue(stream, value.toList())
            }
            is Size -> {
                stream.write(160)
                writeValue(stream, value.toList())
            }
            is SourceQueryOptions -> {
                stream.write(161)
                writeValue(stream, value.toList())
            }
            is StyleObjectInfo -> {
                stream.write(162)
                writeValue(stream, value.toList())
            }
            is StyleProjection -> {
                stream.write(163)
                writeValue(stream, value.toList())
            }
            is StylePropertyValue -> {
                stream.write(164)
                writeValue(stream, value.toList())
            }
            is TileCacheBudgetInMegabytes -> {
                stream.write(165)
                writeValue(stream, value.toList())
            }
            is TileCacheBudgetInTiles -> {
                stream.write(166)
                writeValue(stream, value.toList())
            }
            is TransitionOptions -> {
                stream.write(167)
                writeValue(stream, value.toList())
            }
            else -> super.writeValue(stream, value)
        }
    }
}